File: scripts/importar-excel.py
import pandas as pd
import sqlite3
import os
from datetime import datetime, date

EXCEL = r"C:\Users\pedro.sa\OneDrive - PME MAQUINAS E EQUIPAMENTOS LTDA\PEDRO\RELAÇÃO_GADO - GERAL - 2026.xlsx"
DB    = r"C:\Users\pedro.sa\fazendas-dona-olga\prisma\rebanho.db"

df = pd.read_excel(EXCEL, sheet_name="Geral", header=6)
df.columns = [str(c).strip() for c in df.columns]

# Colunas reais após leitura com header=6:
# 0:Data Reg., 1:Sexo, 2:Tipo, 3:Espec., 4:Pelagem,
# 5:Idade, 6:Nº, 7:Observações, 8:Peso (@), 9:Ref. (@),
# 10:Cor do brinco, 11:LOTE

cols = list(df.columns)

SEXOS_VALIDOS  = {"F", "M"}
TIPOS_VALIDOS  = {"Nelore", "Comum", "Leiteiro"}
ESPEC_VALIDOS  = {"Bezerra", "Novilha", "Vaca", "Bezerro", "Garrote", "Touro"}

animais = []
for _, row in df.iterrows():
    sexo  = str(row.iloc[1]).strip() if pd.notna(row.iloc[1]) else ""
    tipo  = str(row.iloc[2]).strip() if pd.notna(row.iloc[2]) else ""
    espec = str(row.iloc[3]).strip() if pd.notna(row.iloc[3]) else ""

    if sexo not in SEXOS_VALIDOS:
        continue
    if tipo not in TIPOS_VALIDOS:
        continue
    if espec not in ESPEC_VALIDOS:
        continue

    # Data de registro
    data_reg = row.iloc[0]
    if pd.isna(data_reg):
        continue
    if isinstance(data_reg, (datetime, date, pd.Timestamp)):
        data_reg = pd.Timestamp(data_reg).strftime("%Y-%m-%d %H:%M:%S")
    else:
        try:
            data_reg = pd.to_datetime(str(data_reg)).strftime("%Y-%m-%d %H:%M:%S")
        except:
            continue

    pelagem  = str(row.iloc[4]).strip() if pd.notna(row.iloc[4]) else ""
    numero   = str(row.iloc[6]).strip() if pd.notna(row.iloc[6]) else None
    if numero in ("nan", "-", "", "None"):
        numero = None

    obs      = str(row.iloc[7]).strip() if pd.notna(row.iloc[7]) else None
    if obs in ("nan", "None", ""):
        obs = None

    peso_raw = row.iloc[8]
    peso     = float(peso_raw) if pd.notna(peso_raw) and str(peso_raw).replace(".","").isdigit() else None
    try:
        peso = float(peso_raw) if pd.notna(peso_raw) else None
    except:
        peso = None

    ref_raw  = row.iloc[9]
    try:
        peso_ref = float(ref_raw) if pd.notna(ref_raw) else None
    except:
        peso_ref = None

    cor_brinco = str(row.iloc[10]).strip() if pd.notna(row.iloc[10]) else None
    if cor_brinco in ("nan", "None", ""):
        cor_brinco = None

    lote_raw = row.iloc[11]
    lote     = str(lote_raw).strip() if pd.notna(lote_raw) else None
    if lote in ("nan", "None", ""):
        lote = None

    now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    animais.append((
        data_reg,   # dataRegistro
        "Ativo",    # status
        sexo,
        tipo,
        espec,
        pelagem,
        numero,
        None,       # apelido
        obs,
        peso,
        peso_ref,
        cor_brinco,
        lote,
        "importacao",  # criadoPor
        now,        # criadoEm
        now,        # atualizadoEm
    ))

conn = sqlite3.connect(DB)
cur  = conn.cursor()

cur.execute("SELECT COUNT(*) FROM Animal")
existentes = cur.fetchone()[0]
if existentes > 0:
    print(f"Banco já tem {existentes} animais. Pulando importação.")
    conn.close()
else:
    cur.executemany("""
        INSERT INTO Animal
          (dataRegistro, status, sexo, tipo, espec, pelagem,
           numero, apelido, observacoes, peso, pesoRef,
           corBrinco, lote, criadoPor, criadoEm, atualizadoEm)
        VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
    """, animais)
    conn.commit()
    print(f"Importados {len(animais)} animais com sucesso!")

conn.close()

File: scripts/importar-excel.js
/**
 * Script de importação do Excel para o banco SQLite
 * Uso: node scripts/importar-excel.js
 */

const xlsx = require('xlsx');
const Database = require('better-sqlite3');
const path = require('path');
const fs = require('fs');

const EXCEL_PATH = 'C:/Users/pedro.sa/OneDrive - PME MAQUINAS E EQUIPAMENTOS LTDA/PEDRO/SUBIR APP.xlsx';
const DB_PATH = path.join(__dirname, '../prisma/rebanho.db');

// Converte serial do Excel para data ISO
function excelDateToISO(serial) {
  if (!serial || isNaN(serial)) return new Date().toISOString();
  const jsDate = new Date(Math.round((serial - 25569) * 86400 * 1000));
  return jsDate.toISOString();
}

// Limpa valor: "-" ou undefined ou vazio → null
function limpa(v) {
  if (v === undefined || v === null) return null;
  const s = String(v).trim();
  if (s === '-' || s === '' || s.toLowerCase() === 'null') return null;
  return s;
}

// Cor do brinco: "S/Brinco" → null
function limpaCorBrinco(v) {
  const s = limpa(v);
  if (!s || s === 'S/Brinco') return null;
  return s;
}

// Número do animal: pode ser inteiro → string
function limpaNumero(v) {
  if (v === undefined || v === null) return null;
  const s = String(v).trim();
  if (s === '-' || s === '') return null;
  // Se for numérico puro, pode deixar como string
  return s;
}

function main() {
  console.log('📖 Lendo Excel...');
  const wb = xlsx.readFile(EXCEL_PATH);
  const ws = wb.Sheets[wb.SheetNames[0]];
  const data = xlsx.utils.sheet_to_json(ws, { header: 1 });
  const rows = data.slice(1).filter(r => r && r.length > 0);

  console.log(`✅ ${rows.length} linhas encontradas`);

  console.log('🔌 Abrindo banco de dados...');
  const db = new Database(DB_PATH);

  // Limpa dados existentes (mantém auditoria)
  console.log('🗑️  Limpando animais e procedimentos existentes...');
  db.exec('DELETE FROM Procedimento');
  db.exec('DELETE FROM Animal');
  // Reset autoincrement
  db.exec("DELETE FROM sqlite_sequence WHERE name='Animal'");
  db.exec("DELETE FROM sqlite_sequence WHERE name='Procedimento'");

  // Prepara insert
  const insert = db.prepare(`
    INSERT INTO Animal (
      numero, apelido, sexo, tipo, espec, pelagem,
      lote, status, peso, pesoRef,
      corBrinco, observacoes,
      dataRegistro, criadoEm, atualizadoEm,
      numeroMae, apelidoMae, criadoPor
    ) VALUES (
      @numero, @apelido, @sexo, @tipo, @espec, @pelagem,
      @lote, @status, @peso, @pesoRef,
      @corBrinco, @observacoes,
      @dataRegistro, @criadoEm, @atualizadoEm,
      @numeroMae, @apelidoMae, @criadoPor
    )
  `);

  let ok = 0, erros = 0;

  const insertMany = db.transaction((rows) => {
    for (const row of rows) {
      try {
        // Colunas:
        // 0: TIPO ENTRADA  1: STATUS      2: Data Reg.   3: ID REG
        // 4: Sexo          5: ORIGEM      6: Tipo        7: Espec.
        // 8: Pelagem       9: Idade       10: Nº         11: Apelido
        // 12: Nº Mãe      13: Apelido Mãe 14: Observações 15: Peso(KGs)
        // 16: Cor brinco  17: LOTE

        const status  = limpa(row[1]) || 'Ativo';
        const dataReg = excelDateToISO(row[2]);
        const sexo    = limpa(row[4]) || 'F';
        const origem  = limpa(row[5]);
        const tipo    = limpa(row[6]) || 'Nelore';
        const espec   = limpa(row[7]) || 'Vaca';
        const pelagem = limpa(row[8]) || 'Branco';
        const numero  = limpaNumero(row[10]);
        const apelido = limpa(row[11]);
        const numMae  = limpaNumero(row[12]);
        const apMae   = limpa(row[13]);
        const obs     = limpa(row[14]);
        const pesoKg  = row[15] ? parseFloat(row[15]) : null;
        const pesoArr = pesoKg ? pesoKg / 30 : null; // converte para arrobas (1@ = 30kg, rendimento 50%)
        const corBr   = limpaCorBrinco(row[16]);
        const lote    = limpa(row[17]);

        // Monta observações finais
        let observacoes = obs || null;
        if (origem) {
          observacoes = observacoes ? `${observacoes} | Origem: ${origem}` : `Origem: ${origem}`;
        }

        insert.run({
          numero,
          apelido,
          sexo,
          tipo,
          espec,
          pelagem,
          lote,
          status,
          peso: pesoArr,
          pesoRef: pesoArr,
          corBrinco: corBr,
          observacoes,
          dataRegistro: dataReg,
          criadoEm: new Date().toISOString(),
          numeroMae: numMae,
          apelidoMae: apMae,
          atualizadoEm: new Date().toISOString(),
          criadoPor: 'importacao-excel',
        });

        ok++;
      } catch (e) {
        erros++;
        console.error(`❌ Erro na linha:`, row[10], '-', e.message);
      }
    }
  });

  insertMany(rows);

  const total = db.prepare('SELECT COUNT(*) as n FROM Animal').get();
  console.log(`\n✅ Importação concluída!`);
  console.log(`   ✔ ${ok} animais importados`);
  if (erros > 0) console.log(`   ✖ ${erros} erros`);
  console.log(`   📊 Total no banco: ${total.n} animais`);

  // Resumo por status e lote
  const porLote = db.prepare('SELECT lote, COUNT(*) as n FROM Animal GROUP BY lote').all();
  console.log('\n📋 Por lote:');
  porLote.forEach(l => console.log(`   ${l.lote || 'S/Lote'}: ${l.n}`));

  db.close();
}

main();
